core/src/main/scala/org/apache/spark/sql/SnappyContext.scala

```diff
@@ -544,41 +543,108 @@ class SnappyContext protected[spark](
     }
   }
 
-
   /**
-   * Create Index on a SnappyData table (created by a call to createTable).
-   * @todo how can the user invoke this? sql?
-   */
-  private[sql] def createIndexOnTable(tableIdent: QualifiedTableName,
-      sql: String): Unit = {
-    //println("create-index" + " tablename=" + tableName    + " ,sql=" + sql)
+    * Create an index on a table.
+    * @param indexName Index name which goes in the catalog
+    * @param baseTable Fully qualified name of table on which the index is created.
+    * @param indexColumns Columns on which the index has to be created
+    * @param options Options for indexes. For e.g.
+    *                column table index - ("COLOCATE_WITH"->"CUSTOMER").
+    *                row table index - ("INDEX_TYPE"->"GLOBAL HASH") or ("INDEX_TYPE"->"UNIQUE")
+    */
+  def createIndexOnTable(indexName: String,
+                         baseTable: String,
+                         indexColumns: Seq[String],
+                         options: Map[String, String]): Unit = {
 
+    val tableIdent = catalog.newQualifiedTableName(baseTable)
+    val indexIdent = catalog.newQualifiedTableName(indexName)
+
+    if (indexIdent.database != tableIdent.database) {
+      throw new AnalysisException(
+        s"Index and table have different databases " +
+          s"specified ${indexIdent.database} and ${tableIdent.database}")
+    }
     if (!catalog.tableExists(tableIdent)) {
       throw new AnalysisException(
-        s"$tableIdent is not an indexable table")
+        s"Could not find $tableIdent in catalog")
     }
-
-    //println("qualifiedTable=" + qualifiedTable)
+    import scala.collection.JavaConverters._
     catalog.lookupRelation(tableIdent) match {
-      case LogicalRelation(i: IndexableRelation, _) =>
-        i.createIndex(tableIdent.toString, sql)
+      case LogicalRelation(ir: JDBCMutableRelation, _) =>
+        ir.createIndex(indexIdent,
+          tableIdent,
+          indexColumns,
+          options)
+      case LogicalRelation(ir: JDBCAppendableRelation, _) =>
+        ir.createIndex(indexIdent,
+          tableIdent,
+          indexColumns,
+          options)
+        // Main table is updated to store the index information in it. We could have instead used
+        // createIndex method of HiveClient to do this. But, there were two main issues:
+        // a. Schema needs to be added to the HiveTable to supply indexedCols while creating
+        // the index. But, when schema is added to the HiveTable object, it throws a foreign
+        // key constraint failure on the serdes table. Need to look into this.
+        // b. The bigger issue is that the createIndex needs an indexTable for creating this
+        // index. Also, there are multiple things (like implementing HiveIndexHandler)
+        // that are hive specific and can create issues for us from maintenance perspective
+        catalog.alterTableToAddIndexProp(tableIdent, getIndexTable(indexIdent).toString)
+
```

Comment:
Why isn't this encapsulated in JDBCAppendableRelation.createIndex. Then the code here would be:
```
case LogicalRelation(i: IndexableRelation, _) =>
        i.createIndex(indexIdent, tableIdent, indexColumns, options)
```

like before